packages/notifi-react-card/lib/hooks/useIntercomChat.ts

```diff
@@ -1,32 +1,148 @@
+import { ConversationMessagesEntry } from '@notifi-network/notifi-core/dist/models/ConversationMessages';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { ListRange } from 'react-virtuoso';
 
 import { useNotifiClientContext } from '../context';
+import { formatConversationDateTimestamp } from '../utils/datetimeUtils';
+
+export type ChatConversation = FeedEntry &
+  Readonly<{
+    feed: FeedEntry[];
+    createdDate: string | undefined;
+    lastMessage: ChatMessage | undefined;
+  }>;
+
+export type ChatMessage = Readonly<{
+  id: string;
+  message: string;
+  userId: string;
+  createdDate: string;
+  updatedDate: string;
+  conversationId: string;
+}>;
+
+export type MessageDirection = 'INCOMING' | 'OUTGOING';
+
+export type MessagesBlockFeedEntry = Readonly<{
+  type: 'MESSAGES_BLOCK';
+  direction: MessageDirection;
+  messages: ChatMessage[];
+}>;
+
+export type FeedEntry = {
+  id: string;
+  timestamp: string;
+} & MessagesBlockFeedEntry;
 
 type GetConversationMessagesInputProps = Readonly<{
   first?: number;
-  after?: string;
   conversationId: string;
 }>;
 
-export const useConversationMessages = ({
-  first,
-  after,
+export const useIntercomChat = ({
+  first = 5,
   conversationId,
 }: GetConversationMessagesInputProps) => {
+  const [chatMessages, setChatMessages] = useState<ConversationMessagesEntry[]>(
+    [],
+  );
+  const [endCursor, setEndCursor] = useState<string | undefined>();
+  const [hasNextPage, setHasNextPage] = useState<boolean | null>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [atTop, setAtTop] = useState<boolean>(false);
+  const [visibleRange, setVisibleRange] = useState<ListRange>({
+    startIndex: 0,
+    endIndex: 0,
+  });
+  const [isScrolling, setIsScrolling] = useState<boolean | null>();
   const { client } = useNotifiClientContext();
 
-  const input = {
-    first: first,
-    after: after,
-    getConversationMessagesInput: { conversationId },
-  };
+  const getConversationMessages = useCallback(() => {
+    setIsLoading(true);
+    client
+      .getConversationMessages({
+        first,
+        after: endCursor,
+        getConversationMessagesInput: { conversationId },
+      })
+      .then((response) => {
+        if (Array.isArray(response.nodes)) {
+          setChatMessages([...chatMessages, ...response.nodes]);
+        }
+        if (response.pageInfo) {
+          setEndCursor(response.pageInfo.endCursor);
+          setHasNextPage(response.pageInfo.hasNextPage);
+        }
+        setIsLoading(false);
+      });
+  }, [conversationId, endCursor, chatMessages]);
 
-  const getConversationMessages = useCallback(async () => {
-    const response = await client.getConversationMessages(input);
-    return response;
+  // initialization - load first batch of messages
+  useEffect(() => {
+    getConversationMessages();
   }, []);
 
+  // load more messages when scrolling to the top
+  useEffect(() => {
```

Comment:
Ah, I think that you are already appending the messages as the user scrolls to `ChatMessages` state which is an array, correct? If they scroll up, they should be able to render what is in `ChatMessages`? 

I am not sure if need to call `getConversationMessages` again as the user scrolls up.

Reply:
We need to call getConversationMessages each time the user scrolls up to fetch more message data.

Reply:
if it scrolls up, then we will get the data. if it scrolls down, then we grab the data